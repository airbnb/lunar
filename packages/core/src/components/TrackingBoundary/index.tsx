/* eslint-disable no-param-reassign */

import React from 'react';
import componentName from '../../prop-types/componentName';

export type Props = {
  /** Content to wrap. */
  children: NonNullable<React.ReactNode>;
  /** A unique name to identify this boundary in the context stack. */
  name?: string;
};

/** Wraps a component and provides a context stack for click and keydown tracking purposes. */
export default class TrackingBoundary extends React.Component<Props> {
  static propTypes = {
    name: componentName,
  };

  static defaultProps = {
    name: '',
  };

  private handleTrackContext = ({ nativeEvent }: { nativeEvent: MouseEvent | KeyboardEvent }) => {
    const { name } = this.props;

    if (!name) {
      return;
    }

    if (nativeEvent.trackingContext) {
      nativeEvent.trackingContext.push(name);
    } else {
      nativeEvent.trackingContext = [name];
    }
  };

  render() {
    const { children, name } = this.props;

    if (!name) {
      return children;
    }

    return (
      <tracking-boundary
        onClick={this.handleTrackContext}
        onKeyDown={this.handleTrackContext}
        data-tracking-name={name}
      >
        {children}
      </tracking-boundary>
    );
  }
}
