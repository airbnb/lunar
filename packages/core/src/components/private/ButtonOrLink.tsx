import React from 'react';
import removeFocusOnMouseUp from '../../utils/removeFocusOnMouseUp';
import IconAffix from './IconAffix';

export type ButtonOrLinkTypes = HTMLAnchorElement | HTMLButtonElement;

export type Props = {
  [prop: string]: any;
  /** Icon to display at the end the content. */
  afterIcon?: React.ReactNode;
  /** Icon to display at the start the content. */
  beforeIcon?: React.ReactNode;
  /** Content within the button or link. */
  children: NonNullable<React.ReactNode>;
  /** Whether the element is disabled. */
  disabled?: boolean;
  /** Apply flexbox styles to icons. */
  flexAlign?: boolean;
  /** Render as an anchor link with a URL. */
  href?: string;
  /** Whether the element is loading. */
  loading?: boolean;
  /** Callback fired when the element is clicked. */
  onClick?: (event: React.MouseEvent<ButtonOrLinkTypes>) => void;
  /** Callback fired when the element is released. */
  onMouseUp?: (event: React.MouseEvent<ButtonOrLinkTypes>) => void;
  /** When a link, open the target in a new window. */
  openInNewWindow?: boolean;
  /** Rel attribute override for if the component has an href */
  rel?: string;
  /** When a button, the type of button. */
  type?: 'button' | 'submit' | 'reset';
};

export default class ButtonOrLink extends React.Component<Props> {
  static defaultProps = {
    afterIcon: null,
    beforeIcon: null,
    disabled: false,
    flexAlign: false,
    href: '',
    loading: false,
    openInNewWindow: false,
    type: 'button',
  };

  // Intercept click to handle disabled state
  private handleClick = (event: React.MouseEvent<ButtonOrLinkTypes>) => {
    if (this.props.disabled) {
      event.preventDefault();
    } else if (this.props.onClick) {
      this.props.onClick(event);
    }
  };

  // Remove outline after click
  private handleMouseUp = (event: React.MouseEvent<ButtonOrLinkTypes>) => {
    removeFocusOnMouseUp(event);

    if (this.props.onMouseUp) {
      this.props.onMouseUp(event);
    }
  };

  render() {
    const {
      afterIcon,
      beforeIcon,
      children,
      disabled,
      flexAlign,
      href,
      loading,
      openInNewWindow,
      rel,
      type,
      ...restProps
    } = this.props;
    const Tag = href ? 'a' : 'button';
    const props: any = {};

    // Determine props based on element type
    if (href) {
      props.href = href;
      props.rel = rel;

      if (openInNewWindow) {
        props.target = '_blank';

        if (props.rel === undefined) {
          props.rel = 'noopener noreferrer';
        }
      }
    } else {
      props.disabled = disabled || loading || false;
      props.type = type || 'button';
    }

    return (
      // @ts-ignore [ts] JSX element type 'Component' does not have any construct or call signatures. [2604]
      <Tag {...restProps} {...props} onClick={this.handleClick} onMouseUp={this.handleMouseUp}>
        {!loading && beforeIcon && (
          <IconAffix before flex={flexAlign}>
            {beforeIcon}
          </IconAffix>
        )}

        <span>{children}</span>

        {!loading && afterIcon && (
          <IconAffix after flex={flexAlign}>
            {afterIcon}
          </IconAffix>
        )}
      </Tag>
    );
  }
}
