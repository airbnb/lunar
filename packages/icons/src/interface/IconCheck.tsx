import React from 'react';
import withIcon, { Props } from '../withIcon';

function IconCheck(props: Props) {
  return (
    <svg viewBox="0 0 24 24" {...props}>
      <path d="M8.795 15.875l-3.47-3.47a.996.996 0 10-1.41 1.41l4.18 4.18c.39.39 1.02.39 1.41 0l10.58-10.58a.996.996 0 10-1.41-1.41z" />
    </svg>
  );
}

export default withIcon('IconCheck')(IconCheck);
