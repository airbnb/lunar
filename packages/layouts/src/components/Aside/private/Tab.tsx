import React from 'react';
import useStyles, { StyleSheet } from '@airbnb/lunar/lib/hooks/useStyles';
import DirectionalIcon from '@airbnb/lunar/lib/components/DirectionalIcon';
import T from '@airbnb/lunar/lib/components/Translate';
import IconChevronLeft from '@airbnb/lunar-icons/lib/interface/IconChevronLeft';
import IconChevronRight from '@airbnb/lunar-icons/lib/interface/IconChevronRight';

const ICON_SIZE = 18;

const styleSheet: StyleSheet = ({ color, pattern, transition, ui, unit }) => ({
  tab: {
    ...pattern.regularButton,
    ...transition.box,
    border: ui.border,
    borderColor: color.accent.bg,
    backgroundColor: color.accent.bg,
    color: color.core.neutral[5],
    cursor: 'pointer',
    position: 'absolute',
    padding: `${unit}px 0`,
    top: unit * 2,
    transform: 'translate3d(0,0,0)',
    zIndex: 10,
    overflow: 'hidden',

    ':hover': {
      background: color.accent.bgHover,
      color: color.core.neutral[4],
    },
  },

  tab_bordered: {
    borderColor: color.accent.border,
    borderRadius: ui.borderRadius,
  },

  tab_after: {
    right: -ICON_SIZE - ui.borderWidth * 2,
    borderLeftColor: color.accent.bg,
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,

    ':hover': {
      borderLeftColor: color.accent.bgHover,
    },
  },

  tab_before: {
    left: -ICON_SIZE - ui.borderWidth * 2,
    borderRightColor: color.accent.bg,
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,

    ':hover': {
      borderRightColor: color.accent.bgHover,
    },
  },
});

export type Props = {
  /** Whether to apply bordered styles. */
  bordered?: boolean;
  /** Whether the column is expanded. */
  expanded?: boolean;
  /**
   Tab's position relative to the column:
    `after` tab is on the right for a left column (default);
    `before` tab is on the left for a right column.
  */
  position?: 'before' | 'after';
  /** Callback when column is toggled. */
  onCollapseToggle: React.DOMAttributes<HTMLButtonElement>['onClick'];
};

/** A tab for toggling an aside. */
export default function Tab({ bordered, expanded, position = 'after', onCollapseToggle }: Props) {
  const [styles, cx] = useStyles(styleSheet);

  const labelCollapse = T.phrase('Collapse', null, {
    context: 'Collapse',
    key: 'lunar.common.collapse',
  });

  const labelExpand = T.phrase('Expand', null, {
    context: 'Expand',
    key: 'lunar.common.expand',
  });

  let direction: 'left' | 'right' = expanded ? 'left' : 'right';
  if (position === 'before') {
    direction = expanded ? 'right' : 'left';
  }

  return (
    <button
      className={cx(
        styles.tab,
        bordered && styles.tab_bordered,
        position === 'after' && styles.tab_after,
        position === 'before' && styles.tab_before,
      )}
      type="button"
      onClick={onCollapseToggle}
    >
      <DirectionalIcon
        accessibilityLabel={expanded ? labelCollapse : labelExpand}
        direction={direction}
        left={IconChevronLeft}
        right={IconChevronRight}
        size={ICON_SIZE}
      />
    </button>
  );
}
