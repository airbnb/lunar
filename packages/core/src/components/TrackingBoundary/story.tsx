import React from 'react';
import { action } from '@storybook/addon-actions';
import LoremIpsum from ':storybook/components/LoremIpsum';
import TrackingBoundary from '.';

export default {
  title: 'Core/TrackingBoundary',
  parameters: {
    inspectComponents: [TrackingBoundary],
  },
};

export function trackClicksAndKeypressesInAnArea() {
  return (
    <TrackingBoundary name="ReservationsTable">
      {/* eslint-disable-next-line */}
      <div onClick={action('onClick')} onKeyDown={action('onKeyDown')}>
        <LoremIpsum />
      </div>
    </TrackingBoundary>
  );
}

trackClicksAndKeypressesInAnArea.story = {
  name: 'Track clicks and keypresses in an area.',
};
