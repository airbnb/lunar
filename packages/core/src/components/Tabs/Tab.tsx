import React from 'react';
import camelCase from 'lodash/camelCase';
import upperFirst from 'lodash/upperFirst';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ButtonOrLink, { Props as ButtonOrLinkProps } from '../private/ButtonOrLink';
import TrackingBoundary from '../TrackingBoundary';

export type Props = Pick<ButtonOrLinkProps, 'afterIcon' | 'beforeIcon' | 'disabled' | 'href'> & {
  /** Hide bottom border of Tab when unselected. */
  borderless?: boolean;
  /** Content to render if the tab is selected. */
  children?: React.ReactNode;
  /**
   * Unique key name for this tab.
   * @ignore
   */
  keyName?: string;
  /** Text to render in the tab. */
  label: NonNullable<React.ReactNode>;
  /** Callback fired when the tab is clicked. */
  onClick?: (key: string) => void;
  /** Callback fired when the tab is selected. */
  onSelected?: () => void;
  /** Whether the tab is selected or not. */
  selected?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
  /** Stretch tabs to fill the full width. */
  stretched?: boolean;
};

/** A single tab button. Usually rendered amongst a collection of tabs. */
export class Tab extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    borderless: false,
    children: null,
    selected: false,
    small: false,
    stretched: false,
  };

  componentDidUpdate(prevProps: Props) {
    if (!prevProps.selected && this.props.selected && this.props.onSelected) {
      this.props.onSelected();
    }
  }

  private handleClick = () => {
    if (this.props.onClick) {
      this.props.onClick(this.props.keyName!);
    }
  };

  render() {
    const {
      cx,
      afterIcon,
      beforeIcon,
      borderless,
      disabled,
      href,
      keyName,
      label,
      selected,
      small,
      stretched,
      styles,
    } = this.props;
    const trackingName = upperFirst(camelCase(keyName || 'Tab'));

    return (
      <span
        className={cx(
          styles.tab,
          disabled && styles.tab_disabled,
          borderless && styles.tab_borderless,
          selected && styles.tab_selected,
          stretched && styles.tab_stretched,
        )}
      >
        <TrackingBoundary name={trackingName}>
          <ButtonOrLink
            flexAlign
            aria-selected={selected}
            afterIcon={afterIcon}
            beforeIcon={beforeIcon}
            disabled={disabled}
            href={href}
            role="tab"
            onClick={disabled ? undefined : this.handleClick}
            className={cx(
              styles.tabButton,
              small && styles.tabButton_small,
              selected && styles.tabButton_active,
            )}
          >
            {label}
          </ButtonOrLink>
        </TrackingBoundary>
      </span>
    );
  }
}

export default withStyles(({ color, font, pattern, unit, ui, transition }) => ({
  tab: {
    borderBottom: ui.borderThick,
    marginRight: unit * 4,
    marginBottom: -2,

    ':hover': {
      borderColor: color.accent.borderHover,
    },

    ':last-child': {
      marginRight: 0,
    },
  },

  tab_borderless: {
    borderColor: color.clear,
  },

  tab_selected: {
    borderColor: color.accent.borderActive,

    ':hover': {
      borderColor: color.accent.borderActive,
    },
  },

  tab_stretched: {
    flex: 1,
  },

  tab_disabled: {
    ...pattern.disabled,
    borderColor: color.accent.border,

    ':hover': {
      borderColor: color.accent.border,
    },
  },

  tabButton: {
    ...font.textReset,
    ...font.textRegular,
    ...transition.box,
    background: color.accent.bg,
    color: color.accent.text,
    display: 'flex',
    alignItems: 'center',
    paddingTop: unit,
    paddingBottom: unit,
    border: 0,
    cursor: 'pointer',
    textAlign: 'left',
    textDecoration: 'none',
    whiteSpace: 'nowrap',
    width: '100%',
    fontWeight: font.weights.bold,
    borderTopLeftRadius: ui.borderRadius,
    borderTopRightRadius: ui.borderRadius,
  },

  tabButton_small: {
    ...font.textSmall,
  },

  tabButton_active: {
    color: color.accent.textActive,
  },
}))(Tab);
