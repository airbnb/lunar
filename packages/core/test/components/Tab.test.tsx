import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import IconCheck from '@airbnb/lunar-icons/lib/interface/IconCheck';
import Tab from '../../src/components/Tabs/Tab';
import TrackingBoundary from '../../src/components/TrackingBoundary';
import ButtonOrLink from '../../src/components/private/ButtonOrLink';

describe('<Tab/>', () => {
  it('renders a button', () => {
    const wrapper = shallowWithStyles(<Tab keyName="default" label="Tab" onClick={() => {}} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a before icon', () => {
    const icon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}} disabled beforeIcon={icon} />,
    );

    expect(wrapper.find(ButtonOrLink).prop('beforeIcon')).toBe(icon);
  });

  it('renders a after icon', () => {
    const icon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}} disabled afterIcon={icon} />,
    );

    expect(wrapper.find(ButtonOrLink).prop('afterIcon')).toBe(icon);
  });

  it('renders disabled', () => {
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}} disabled />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders selected', () => {
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}} selected />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders stretched', () => {
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}} stretched />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('doesnt render children', () => {
    const child = <div>Foo</div>;
    const wrapper = shallowWithStyles(
      <Tab keyName="default" label="Tab" onClick={() => {}}>
        {child}
      </Tab>,
    );

    expect(wrapper.contains(child)).toBe(false);
  });

  it('triggers `onClick` when clicked', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(<Tab keyName="foo" label="Tab" onClick={spy} />);
    wrapper.findWhere(child => child.prop('onClick')).simulate('click');

    expect(spy).toHaveBeenCalledWith('foo');
  });

  it('triggers `onSelected` when selected', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(
      <Tab keyName="foo" label="Tab" onClick={() => {}} onSelected={spy} />,
    );

    expect(spy).not.toHaveBeenCalled();

    wrapper.setProps({
      selected: true,
    });

    expect(spy).toHaveBeenCalledWith();
  });

  it('renders TrackingBoundary with formatted keyName', () => {
    const wrapper = shallowWithStyles(<Tab keyName="fooName" label="Tab" />);

    expect(wrapper.find(TrackingBoundary).prop('name')).toMatch(/FooName/);
  });

  it('defaults trackingName to Tab if keyName is blank', () => {
    const wrapper = shallowWithStyles(<Tab label="Tab" />);

    expect(wrapper.find(TrackingBoundary).prop('name')).toMatch('Tab');
  });
});
