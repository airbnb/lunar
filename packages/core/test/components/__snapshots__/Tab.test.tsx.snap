// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab/> renders a button 1`] = `
<span
  className="tab_oe67rv"
>
  <TrackingBoundary
    name="Default"
  >
    <ButtonOrLink
      aria-selected={false}
      className="tabButton_17c4m1l"
      onClick={[Function]}
      role="tab"
    >
      Tab
    </ButtonOrLink>
  </TrackingBoundary>
</span>
`;

exports[`<Tab/> renders disabled 1`] = `
<span
  className="tab_oe67rv-o_O-tab_disabled_12oo6iq"
>
  <TrackingBoundary
    name="Default"
  >
    <ButtonOrLink
      aria-selected={false}
      className="tabButton_17c4m1l"
      disabled={true}
      role="tab"
    >
      Tab
    </ButtonOrLink>
  </TrackingBoundary>
</span>
`;

exports[`<Tab/> renders selected 1`] = `
<span
  className="tab_oe67rv-o_O-tab_selected_dfv9v3"
>
  <TrackingBoundary
    name="Default"
  >
    <ButtonOrLink
      aria-selected={true}
      className="tabButton_17c4m1l"
      onClick={[Function]}
      role="tab"
    >
      Tab
    </ButtonOrLink>
  </TrackingBoundary>
</span>
`;

exports[`<Tab/> renders stretched 1`] = `
<span
  className="tab_oe67rv-o_O-tab_stretched_bc4egv"
>
  <TrackingBoundary
    name="Default"
  >
    <ButtonOrLink
      aria-selected={false}
      className="tabButton_17c4m1l"
      onClick={[Function]}
      role="tab"
    >
      Tab
    </ButtonOrLink>
  </TrackingBoundary>
</span>
`;
